core/services/workflows/syncer/v2/workflow_registry.go

```diff
@@ -189,20 +193,32 @@ func (w *workflowRegistry) Start(_ context.Context) error {
 			reader, err := w.newWorkflowRegistryContractReader(ctx)
 			if err != nil {
 				w.lggr.Criticalf("contract reader unavailable : %s", err)
+				w.hooks.OnStartFailure(err)
+				cancel()
 				return
 			}
 
 			w.contractReader = reader
+			close(initDoneCh)
+			w.lggr.Debugw("Received DON and set ContractReader")
 		}()
 
 		w.wg.Add(1)
 		go func() {
 			defer w.wg.Done()
 			defer cancel()
 			// Start goroutines to gather changes from Workflow Registry contract
-			<-initDoneCh
+			select {
+			case <-initDoneCh:
+			case <-ctx.Done():
+				return
+			}
 			w.lggr.Debugw("read from don received channel while waiting to start reconciliation sync")
-			don, _ := w.workflowDonNotifier.WaitForDon(ctx)
+			don, err := w.workflowDonNotifier.WaitForDon(ctx)
```

Comment:
should never happen, but handling error now by passing it to client provided hook with default noop implementation